extensions/stats/config.h

```diff
@@ -0,0 +1,384 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <google/protobuf/util/json_util.h>
+#include "absl/container/flat_hash_map.h"
+#include "extensions/common/context.h"
+#include "extensions/common/node_info.pb.h"
+#include "extensions/stats/config.pb.h"
+
+// WASM_PROLOG
+#ifndef NULL_PLUGIN
+#include "api/wasm/cpp/proxy_wasm_intrinsics.h"
+
+#else  // NULL_PLUGIN
+
+#include "extensions/common/wasm/null/null.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+namespace Null {
+namespace Plugin {
+#endif  // NULL_PLUGIN
+
+// END WASM_PROLOG
+
+namespace Stats {
+
+using StringView = absl::string_view;
+
+constexpr StringView Sep = "#";
+
+// The following need to be std::strings because the receiver expects a string.
+const std::string unknown = "unknown";
+const std::string vSource = "source";
+const std::string vDest = "destination";
+const std::string vMTLS = "mutual_tls";
+const std::string vNone = "none";
+const std::string vDash = "-";
+
+using google::protobuf::util::JsonParseOptions;
+using google::protobuf::util::Status;
+
+struct Node {
+  common::NodeInfo node_info;
+  // key computed from the
+  std::string key;
+
+  Node(common::NodeInfo nodeInfo) {
+    auto labels = nodeInfo.labels();
+    absl::StrAppend(&key, nodeInfo.workload_name(), Sep, nodeInfo.namespace_(),
+                    Sep, labels["app"], Sep, labels["version"]);
+    node_info = nodeInfo;
+  }
+
+  Node() = delete;
+};
+
+using NodeSharedPtr = std::shared_ptr<Node>;
+
+class NodeInfoCache {
+ public:
+  // Fetches and caches Peer information by peerId
+  // TODO Remove this when it is cheap to directly get things from StreamInfo.
+  // At present this involves de-serializing to google.Protobuf.Struct and then
+  // another round trip to NodeInfo. This Should at most hold N entries.
+  const NodeSharedPtr getPeerById(StringView peerMetadataIdKey,
+                                  StringView peerMetadataKey);
+
+ private:
+  absl::flat_hash_map<std::string, NodeSharedPtr> cache_;
+};
+
+using valueExtractorFn = uint64_t (*)(const Common::RequestInfo& request_info);
+
+class SimpleStat {
+ public:
+  SimpleStat(uint32_t metric_id, valueExtractorFn value)
+      : metric_id_(metric_id), valueFn_(value){};
+
+  inline void record(const Common::RequestInfo& request_info) {
+    recordMetric(metric_id_, valueFn_(request_info));
+  };
+
+ private:
+  uint32_t metric_id_;
+  valueExtractorFn valueFn_;
+};
+
+using SimpleStatSharedPtr = std::shared_ptr<SimpleStat>;
```

Comment:
A SimpleStat is copy-able so you don't need to use a shared_ptr.   Just treat it as an immediate and don't worry about lifetime.